### PR TITLE
Use web worker to throttle counts and schema changes

### PR DESCRIPTION
**Pull request by nwmac**
This PR adds a web worker to the cluster store that:

- Intercepts and buffers count change messages and only sends at most 1 every 5 seconds
- Intercepts schema change messages and checks that the schema has actually changed (via a simple hash) and only propagates the change if the schema has changed

The schema change logic ensures that the Sortable Table 'headers' property does not keep changing due to the schema change messages being processed when there is no actual change.

Note that with this PR, we don't need https://github.com/rancher/dashboard/pull/6389

### CODE REVIEW

**Review by aalves08 — APPROVED** [review]
Clever piece of code here @nwmac ! congrats! A bit lost on some architecture params/functions of the dashboard-store but it's more that enough to get the overall picture. LGTM! 💪 we can also play with the debounce times if we ever hit any bottleneck with accumulated queued changes to schema, but I think they will rarely update anyway so, that's a non-issue imo.

Looking forward to see the performance gains on this 👁️

File: shell/plugins/steve/worker.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    const hash = hashObj(schema);
+    const existing = state.schemas[schema.id];
+
+    if (existing && existing !== hash) {
+      // console.log(`${ schema.id } CHANGED  ${ hash } > ${ existing }`);
+      state.schemas[schema.id] = hash;
+
+      const msg = {
+        data:          schema,
+        resoourceType: 'schema',
```
Comment (richard-cox) [inline shell/plugins/steve/worker.js:47]: `resoourceType` typo

File: shell/plugins/steve/subscribe.js [hunk 2]
```diff
@@ -524,10 +593,19 @@ export const actions = {
   },
 
   'ws.resource.remove'(ctx, msg) {
-    queueChange(ctx, msg, false, 'Remove');
-
     const data = msg.data;
     const type = data.type;
+
+    if (type === 'schema') {
```
Comment (richard-cox) [inline shell/plugins/steve/subscribe.js:599]: References to `schema` need a bit of `import { SCHEMA } from '@shell/config/types';` lovin'

**Review by richard-cox — CHANGES_REQUESTED** [review]
I couldn't quite get this to update an existing schema. 
- I fixed the `resoourceType` typo
- Added below to `flush` (before hash)
  ```
      if (schema.id === 'management.cattle.io.token') {
      schema.randomString = new Date().toUTCString();
    }
  ```
On the honker of an instance it triggered the `load` in the worker's flush but i couldn't see the new property on page when looking at `this.$store.getters[`cluster/schemaFor`](SCHEMA)`

I ran out of time to debug, but it looks like the subscribe queueChange is `return`ing early  given a missing `entry`.  There might be an arg's issue

File: shell/plugins/steve/worker.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  return new Uint32Array([hash])[0].toString(36);
+}
+
+// Quick, simple hash function to generate hash for an object
+function hashObj(obj) {
+  return hash(JSON.stringify(obj, null, 2));
+}
+
+function flush() {
+  state.queue.forEach((item) => {
```
[the remark this answers is not in the document]
Comment (nwmac) [reply shell/plugins/steve/worker.js:36]: Yep, still single threaded

File: shell/plugins/steve/worker.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    close();
+  },
+
+  // Debounce counts messages so we only process at most 1 every 5 seconds
+  countsUpdate(resource) {
+    state.counts.push(resource);
+
+    if (!state.countTimer) {
+      state.countTimer = setTimeout(() => {
+        const last = state.counts[state.counts.length - 1];
```
[the remark this answers is not in the document]
Comment (nwmac) [reply shell/plugins/steve/worker.js:90]: Done

File: shell/plugins/steve/worker.js (hunk 1 above)
Comment (nwmac) [reply shell/plugins/steve/worker.js:47]: Done - don't think this is used - I think code looks at `data.type`

File: shell/plugins/steve/index.js [hunk 5]
```diff
@@ -64,7 +64,10 @@ export default (config) => {
   }
 
   return coreStore(
-    SteveFactory(config),
+    SteveFactory(config.namespace, config.baseUrl),
```
[the remark this answers is not in the document]
Comment (nwmac) [reply shell/plugins/steve/index.js:67]: Indeed

**Comment by nwmac** [comment]
@richard-cox Updated

Your way of testing is wired into the schema update flush - so you'd only get that if that schema changed - that would cause the flush and you'd add that new property.

I tested with this:

Add this code into `shell/plugins/steve/mutations.js`, into the `loadMulti` function in the for loop (before the pod check):

```
      if (resource.type === 'schema') {
        console.warn(`Loaded schema: ${ entry.id }`);
      }
```
This will log when a schema is updated.

Bring up a downstream cluster.

Use the example CRD from here: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/ (assume you've saved this as `crd.yaml`

Add that into your cluster with:

`kubectl apply -f crd.yaml`

Observe that in the console log, you'll get a load for the new schema.

Edit the `crd.yaml` and add another name to the `shortNames` at the end.

Use kubectl to apply this - you'll see websocket messages for schema changes, including this one, but the data coming back will not have changed (the shortnames is not part of this), so you should not see the console log message.

Delete the CRD with:

`kubectl delete crdcrontabs.stable.example.com`

Now add the CRD again, and you should get the console log message again.

**Review by richard-cox — APPROVED** [review]

**Pull request merged**